### ISSUE TITLE
aleatoric uncertainty loss term

### ISSUE THREAD

**Issue by jin8** [pmorerio/dl-uncertainty#1]
Hi, is there a reason why you did not put activation function for the mu, logvar at the end of the decoder.
I am getting negative values for the loss. Is it okay to have a negative loss for the model??

**Comment by pmorerio** [comment]
Hi. 

Concerning the ``mean`` branch of the decoder, I am not putting any activation because I am generating images  in the range [-1,1] (check the [load_mnist function](https://github.com/pmorerio/dl-uncertainty/blob/8c96c7a2ccf0453a4e0994f7ef1911a955b35b91/aleatoric-uncertainty/solver.py#L30)), thus I want the decoder to generate at least in that range. In principle you could also use ``tanh`` activation for that, but that was giving numerical issues.

Concerning the ``log_var`` branch, you should allow for any real number, (also negative values, when the variance is smaller than 1). This is why, again, no activation should be needed in my opinion.

Eq.(8) in the paper  [arXiv version](https://arxiv.org/pdf/1703.04977.pdf) seems to suggest that a negative loss is allowed. 
![image](https://user-images.githubusercontent.com/6783835/58402605-0ffe0700-8061-11e9-909d-d73c6e3319c5.png)
However the loss has a lower bound (see pic below), provided ||y-\hat{y}|| is not zero. The minimum can be negative.

![image](https://user-images.githubusercontent.com/6783835/58102162-22032400-7be1-11e9-95d9-2a1c1344f81c.png)

However, this is my personal understanding. Please feel free to find any fault in it.

Best,

P.

**Comment by sdsy888** [comment]
Hi @pmorerio , does the variable `log_var` in your [code](https://github.com/pmorerio/dl-uncertainty/blob/2b003a225b68de45d68811c275905bb84103dbe6/aleatoric-uncertainty/model.py#L74) mean `logits_variable` or `log(function) variance`?

 As indicated in the paper, the `variance`  $\delta$ should be processed by `log` function to become  $s_i$: 
<img width="123" alt="WX20190628-025117@2x" src="https://user-images.githubusercontent.com/7611912/60334746-09f59180-9951-11e9-91cf-f2f5b63cc8c2.png">

To my understanding, the `logits` from the conv2d function aren't processed by `log` function.

Is there any reason why you didn't `tf.log` the variance explicitly?  

Thank you!

**Comment by pmorerio** [comment]
Hi @sdsy888, thanks for your question.
The variable `log_var` is precisely  $s_i$, and I named it after the `log` function. The network is regressing $s_i$ directly, instead of predicting $\sigma_i^2$, in order to overcome possible numerical issues given by the `log` function when applied to the variance (from the paper: '_In practice, we train the network to predict the log variance [...] This is because it is more numerically stable than regressing the variance_'). In fact, the [loss](https://github.com/pmorerio/dl-uncertainty/blob/2b003a225b68de45d68811c275905bb84103dbe6/aleatoric-uncertainty/model.py#L74) is then written as in eq (8), with `log_var` in place of $s_i$. 

The variable `log_var` is later on exponentiated in order to [save the actual variance](https://github.com/pmorerio/dl-uncertainty/blob/2b003a225b68de45d68811c275905bb84103dbe6/aleatoric-uncertainty/model.py#L91) in the `tf.summary`.

Hope this clarifies.

**Comment by sdsy888** [comment]
> Hi @sdsy888, thanks for your question.
> The variable `log_var` is precisely $s_i$, and I named it after the `log` function. The network is regressing $s_i$ directly, instead of predicting $\sigma_i^2$, in order to overcome possible numerical issues given by the `log` function when applied to the variance (from the paper: '_In practice, we train the network to predict the log variance [...] This is because it is more numerically stable than regressing the variance_'). In fact, the [loss](https://github.com/pmorerio/dl-uncertainty/blob/2b003a225b68de45d68811c275905bb84103dbe6/aleatoric-uncertainty/model.py#L74) is then written as in eq (8), with `log_var` in place of $s_i$.
> 
> The variable `log_var` is later on exponentiated in order to [save the actual variance](https://github.com/pmorerio/dl-uncertainty/blob/2b003a225b68de45d68811c275905bb84103dbe6/aleatoric-uncertainty/model.py#L91) in the `tf.summary`.
> 
> Hope this clarifies.

Oh, I thought it in the wrong direction. I thought eq.8 will first predict variance and then get the real value using `exp` Thank you!

**Comment by ShellingFord221** [comment]
It seems that in your code of aleatoric uncertainty, mean and log_var are two separate parameters trained in the model. I thought that they should come from the output of the model (i.e. mean means the average of output y, log_var means the log variance of the output y). Why model them as two variables rather than results from y? Thanks!

**Comment by pmorerio** [comment]
Hi, from my understanding what you describe is epistemic uncertainty, practically implemented by sampling different networks via dropout at test time and calculating average and variance. 

To model aleatoric uncertainty  I follow equation 6 in the paper:
![image](https://user-images.githubusercontent.com/6783835/70314207-52c24800-1817-11ea-8c1e-0611d2706029.png)
Both mean and variance are output of the net f ('head is split to predict both mean and covariance'  reads the text below).

Well at least this is my understanding.

**Comment by ShellingFord221** [comment]
Thank you for your answer! I'm a little bit confused that how to 'use a single network to transform the input with its head split to predict both mean as well as variance'? Just replace them with the single output y in the model? Are there any supporting theories? Thanks!

**Comment by pmorerio** [comment]
Sorry I did not really get your questions.

**Comment by ShellingFord221** [comment]
Sorry for the disambiguated description. I mean that in your code, it seems that you first pass the image through each layer of the network, then decode the output by passing it through the same layers of the network to get mean and variance. Why is that?

**Comment by pmorerio** [comment]
Hi, 
the network has encoder-decoder structure, since I want to regress images.
Images goes through the encoder, and one decoding layer. After that the decoder 'splits' the output to have a prediction for the 'mean' and a prediction for the 'log_var'. Both predictions contribute to the loss.
I guess you can 'split' before or after, that would not make much difference. I have no intuition nor supporting theory for my design choice, but I guess you need at least a couple of layers for each branch.

I hope this answers your question.
(For any other question, please open a new issue, so that different topic are not mixed in a single thread)
